### PR TITLE
Verification Task initial PR (w/o data)

### PR DESCRIPTION
**Pull request by albutko**
Wanted to go ahead and have some extra eyes on the verification task code. It is about 95% complete, I just need to add the code to download/create the dataset along with updated `README.md`. I have added comments to all functions in the style of @spongezhang. 

But there are quite a few files that could use some checking:
- `geom.py` : the geometric helper functions (Calculating Epipolar Distances, calculating F, E, etc.)
- `epiConstraintBench.py`: epipolar constraint benchmark
- `epiDistanceBench.py`: epipolar distance benchmark
- `inlierPrecisionBench.py`: Number of inliers, Inlier Percentage, Precision, and Recall counts 
- `matrixNormBench.py`: Frobenius norm of difference from groundtruth Fundamental matrix
- `VerificationTemplate.py`: template for Verification algorithms

I based the verification dataset off of the dataset used in [Learned Correspondences](https://github.com/vcg-uvic/learned-correspondence-release). Because of this, it is slightly different than the sequence datasets. I am happy to reconsider the structure if need be. 

**I will add the data download and extraction code in the next few days within this PR** so, *in case of the small probability that everything looks completely correct from a math stand point*, please don't merge yet. Happy to get any constructive criticism! 

Thanks,
Alex

### CODE REVIEW

File: python/test/test_allVerification_bench.py [hunk 1]
```diff
@@ -0,0 +1,71 @@
+#!/usr/bin/python
+# -*- coding: utf-8 -*-
+# ===========================================================
+#  File Name: test_rep_bench.py
+#  Author: Xu Zhang, Columbia University
```
Comment (johnwlambert) [inline python/test/test_allVerification_bench.py]: @albutko Things are looking good. Probably would be good to change this line to your name, and put Georgia Tech for the files you've written.

File: python/verifiers/ransac.py [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+                estimates_essential=True,
+                estimates_fundamental=True)
+
+    def evaluate_matches(self, pts1, pts2):
+        _, mask = cv2.findEssentialMat(pts1, pts2, method=cv2.RANSAC)
+        inlier_pts1 = pts1[mask.ravel() == 1]
+        inlier_pts2 = pts2[mask.ravel() == 1]
+        return (inlier_pts1, inlier_pts2)
+
+    def estimate_essential_matrix(self, pts1, pts2):
```
Comment (johnwlambert) [inline python/verifiers/ransac.py]: @albutko can we add a docstring with function description and type of arguments here?

File: python/verifiers/ransac.py [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+    def __init__(self):
+        super(
+            RANSAC,
+            self).__init__(
+                name='RANSAC',
+                estimates_essential=True,
+                estimates_fundamental=True)
+
+    def evaluate_matches(self, pts1, pts2):
+        _, mask = cv2.findEssentialMat(pts1, pts2, method=cv2.RANSAC)
```
Comment (johnwlambert) [inline python/verifiers/ransac.py]: @albutko same story here. a good example is https://github.com/argoai/argoverse-api/blob/master/argoverse/utils/cv2_plotting_utils.py#L21

File: python/verifiers/mlesac.py [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+            if sum_err < best_err:
+                best_err = sum_err
+                best_E = E
+                best_inliers = inliers
+
+        best_inliers = best_inliers.reshape(-1, 1).astype(np.uint8)
+
+        return best_E
+
+    def estimate_fundamental_matrix(self, np1, np2):
```
Comment (johnwlambert) [inline python/verifiers/mlesac.py]: function is a bit long currently, can we break parts of it into a few smaller functions? we can make `min_pt_num = 8` a global variable in this file as `MIN_F_MATRIX_PT_NUM = 8`

File: python/verifiers/mlesac.py [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+                best_err = sum_err
+                best_F = F
+                best_inliers = inliers
+
+        best_inliers = best_inliers.reshape(-1, 1).astype(np.uint8)
+
+        return best_F
+
+def compute_error(x1, x2, E):
+    Ex1 = E.dot(x1.T)
```
Comment (johnwlambert) [inline python/verifiers/mlesac.py]: same story here about docstring and the argument types

File: python/verifiers/learnedCorres.py [hunk 6]
```diff
@@ -0,0 +1,684 @@
...
+
+        return res["e_hat"].reshape(3,3)
+
+
+
+class Config:
+    def __init__(self, **kwargs):
+        self.__dict__.update(kwargs)
+
+class MyNetwork(object):
```
Comment (johnwlambert) [inline python/verifiers/learnedCorres.py]: can we name this class something a bit more descriptive?

File: python/verifiers/learnedCorres.py [hunk 7]
```diff
@@ -0,0 +1,684 @@
...
+
+import cv2
+import numpy as np
+import tensorflow as tf
+
+dirname = os.path.dirname(os.path.abspath(__file__))
+
+from verifiers.VerificationTemplate import VerificationTemplate
+
+class learnedCorres(VerificationTemplate):
```
Comment (johnwlambert) [inline python/verifiers/learnedCorres.py]: can we add a comment here with a brief description about what this class is to be used for?

File: python/verifiers/VerificationTemplate.py [hunk 8]
```diff
@@ -0,0 +1,29 @@
+
+
+
+class VerificationTemplate():
```
Comment (johnwlambert) [inline python/verifiers/VerificationTemplate.py]: Very nice work with the abstract class

File: python/test/test_inlierPrec_bench.py [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+    mlesac = verifiers.mlesac.MLESAC()
+
+    # Define dataset
+    dataset = dset.verification_dataset.verification_dataset(['reichstag'])
+
+    # Do the evaluation
+    rep_result_r = inlierPrecisionBench.evaluate(
+        dataset, mlesac, use_cache=False, save_result=True)
+
+    # rep_result_l = inlierPrecisionBench.evaluate(
```
Comment (johnwlambert) [inline python/test/test_inlierPrec_bench.py]: can we remove this commented out code?

File: python/test/test_inlierPrec_bench.py [hunk 10]
```diff
@@ -0,0 +1,65 @@
+#!/usr/bin/python
+# -*- coding: utf-8 -*-
+# ===========================================================
+#  File Name: test_rep_bench.py
+#  Author: Xu Zhang, Columbia University
```
Comment (johnwlambert) [inline python/test/test_inlierPrec_bench.py]: same story about the copyright notice

File: python/test/test_epiDist_bench.py [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+        dataset, lmeds, use_cache=False, save_result=True)
+
+    rep_result = [rep_result_r, rep_result_l]
+
+    # Show the result
+    for result_term in rep_result[0]['result_term_list']:
+        bench.Utils.print_result(rep_result, result_term)
+        bench.Utils.save_result(rep_result, result_term)
+
+    #Show result for different sequences
```
Comment (johnwlambert) [inline python/test/test_epiDist_bench.py]: do we need this commented code? could it be under a `verbose` flag or `save_to_disk` flag, perhaps default variable in function signature?

File: python/dset/verification_dataset.py [hunk 12]
```diff
@@ -0,0 +1,217 @@
...
+                            data[data_name][var_name] += pickle.load(ifp)
+                        else:
+                            data[data_name][var_name] = pickle.load(ifp)
+                    else:
+                        data[data_name] = {}
+                        data[data_name][var_name] = pickle.load(ifp)
+
+        self.data = data
+
+    def _prepare_data(self):
```
Comment (johnwlambert) [inline python/dset/verification_dataset.py]: function is a bit long currently, can we make it shorter? one simple way  would be to combine lines, e.g.

`coords1, coords2 = list(), list()`

File: python/bench/matrixNormBench.py [hunk 13]
```diff
@@ -0,0 +1,84 @@
...
+# -*- coding: utf-8 -*-
+# ===========================================================
+#  File Name: repBench.py
+#  Author: Xu Zhang, Columbia University
+#  Creation Date: 01-25-2019
+#  Last Modified: Tue Mar  5 21:46:25 2019
+#
+#  Description:repeatability benchmark
+#
+#  Copyright (C) 2018 Xu Zhang
```
Comment (johnwlambert) [inline python/bench/matrixNormBench.py]: we can update year to 2019

File: python/bench/geom.py [hunk 14]
```diff
@@ -0,0 +1,645 @@
...
+# Version:
+# Package-Requires: ()
+# URL:
+# Doc URL:
+# Keywords:
+# Compatibility:
+#
+#
+
+# Commentary:
```
Comment (johnwlambert) [inline python/bench/geom.py]: can we remove these empty blank lines with only `#` on them?

File: python/bench/geom.py [hunk 15]
```diff
@@ -0,0 +1,645 @@
...
+    return M
+
+
+def np_unskew_symmetric(M):
+    """
+    From: https://github.com/vcg-uvic/learned-correspondence-release
+
+    Inverse cross product matrix to extract underlying vector v in R^3
+    """
+
```
Comment (johnwlambert) [inline python/bench/geom.py]: can probably remove this blank line

File: python/bench/geom.py [hunk 16]
```diff
@@ -0,0 +1,645 @@
...
+        rotation matrix from img1 to img2
+    dt: np.array (3x1)
+        translation vector of camera1 center (img1) to camera2 center (img2)
+
+    Output
+    ------
+    ys: np.array (N,)
+        ys[idx] is the epipolar symmetric distance of correspondence (x1[idx], x2[idx])
+    """
+
```
Comment (johnwlambert) [inline python/bench/geom.py]: can probably remove blank line here

File: python/bench/epiDistanceBench.py [hunk 17]
```diff
@@ -0,0 +1,104 @@
...
+        est_F = data_dict['est_F']
+        pts1 = data_dict['px_coords1']
+        pts2 = data_dict['px_coords1']
+
+        if est_F is None:
+            est_F = geom.get_F_matrix_from_E(data_dict['est_E'],
+                                             data_dict['K1'],
+                                             data_dict['K2'])
+
+
```
Comment (johnwlambert) [inline python/bench/epiDistanceBench.py]: can remove blank line here

File: python/bench/epiConstraintBench.py [hunk 18]
```diff
@@ -0,0 +1,103 @@
...
+        :returns: result
+        :rtype: dict
+
+        See Also
+        --------
+
+        bench.Benchmark
+        bench.Benchmark.evaluate_warpper:
+        """
+
```
Comment (johnwlambert) [inline python/bench/epiConstraintBench.py]: can probably remove blank line

File: python/bench/allVerificationBenchs.py [hunk 19]
```diff
@@ -0,0 +1,160 @@
+#!/usr/bin/python
+# -*- coding: utf-8 -*-
+# ===========================================================
+#  File Name: repBench.py
+#  Author: Xu Zhang, Columbia University
```
Comment (johnwlambert) [inline python/bench/allVerificationBenchs.py]: wasn't repBench already in the repo? what is this change?

File: python/test/test_allVerification_bench.py (hunk 1 above)
Comment (johnwlambert) [reply python/test/test_allVerification_bench.py]: (if this was a file that you wrote)

File: python/bench/VerificationBenchmarkTemplate.py [hunk 20]
```diff
@@ -0,0 +1,277 @@
...
+    """
+
+    def __init__(self, name,
+                 result_dir='./python_scores/'):
+        self.name = name
+        self.result_dir = result_dir
+
+
+
+    def find_F_matrix(self, verifier, data_dict):
```
Comment (johnwlambert) [inline python/bench/VerificationBenchmarkTemplate.py]: would be cool to add Python typing to help with readability, like here: https://github.com/argoai/argoverse-api/blob/master/argoverse/utils/cv2_plotting_utils.py#L21

File: python/verifiers/learnedCorres.py (hunk 7 above)
Comment (albutko) [reply python/verifiers/learnedCorres.py]: What about the comment I added below?

**Pull request merged**